### ISSUE TITLE
Regression in MachineIndependent: error: 'operator=' is a protected member of 'glslang::pool_allocator<char>

### ISSUE THREAD

**Issue by EoD** [KhronosGroup/glslang#879]
Commit c10191d1329b9a7fe4438e6be60e2f8e2894254c introduces a regression which is not  fixed by 1ee1c8e1ed6d74a85c3824c923e37ace98cd9899 .

With GCC
```
[ 28%] Building CXX object glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/../Include/Common.h:99:0,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54,
                 from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h: In instantiation of ‘void std::swap(_Tp&, _Tp&) [with _Tp = glslang::pool_allocator<char>]’:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/allocator.h:175:8:   required from ‘static void std::__alloc_swap<_Alloc, false>::_S_do_it(_Alloc&, _Alloc&) [with _Alloc = glslang::pool_allocator<char>]’
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.tcc:67:50:   required from ‘void std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::swap(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = glslang::pool_allocator<char>]’
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.h:589:2:   required from ‘std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = glslang::pool_allocator<char>]’
/workspace/glslang/glslang/MachineIndependent/Initialize.cpp:3854:46:   required from here
/workspace/glslang/glslang/MachineIndependent/../Include/../Include/PoolAlloc.h:318:21: error: ‘glslang::pool_allocator<T>& glslang::pool_allocator<T>::operator=(const glslang::pool_allocator<T>&) [with T = char]’ is protected
     pool_allocator& operator=(const pool_allocator&) { return *this; }
                     ^
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_pair.h:59:0,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_algobase.h:64,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/char_traits.h:39,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/ios:40,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/istream:38,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/sstream:38,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/../Include/Common.h:55,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54,
                 from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h:186:11: error: within this context
       __a = _GLIBCXX_MOVE(__b);
           ^
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/../Include/Common.h:99:0,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54,
                 from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
/workspace/glslang/glslang/MachineIndependent/../Include/../Include/PoolAlloc.h:318:21: error: ‘glslang::pool_allocator<T>& glslang::pool_allocator<T>::operator=(const glslang::pool_allocator<T>&) [with T = char]’ is protected
     pool_allocator& operator=(const pool_allocator&) { return *this; }
                     ^
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_pair.h:59:0,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_algobase.h:64,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/char_traits.h:39,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/ios:40,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/istream:38,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/sstream:38,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/../Include/Common.h:55,
                 from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54,
                 from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h:187:11: error: within this context
       __b = _GLIBCXX_MOVE(__tmp);
           ^
make[2]: *** [glslang/CMakeFiles/glslang.dir/build.make:159: glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1004: glslang/CMakeFiles/glslang.dir/all] Error 2
make: *** [Makefile:139: all] Error 2
```

With Clang
```
[ 28%] Building CXX object glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o
In file included from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54:
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/Common.h:55:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/sstream:38:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/istream:38:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/ios:40:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/char_traits.h:39:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_algobase.h:64:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_pair.h:59:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h:186:11: error: 'operator=' is a protected member of 'glslang::pool_allocator<char>'
      __a = _GLIBCXX_MOVE(__b);
      ~~~ ^ ~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/allocator.h:175:4: note: in instantiation of function template specialization 'std::swap<glslang::pool_allocator<char> >' requested here
          swap(__one, __two);
          ^
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.tcc:67:42: note: in instantiation of member function 'std::__alloc_swap<glslang::pool_allocator<char>, false>::_S_do_it' requested here
      std::__alloc_swap<allocator_type>::_S_do_it(_M_get_allocator(),
                                         ^
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.h:589:8: note: in instantiation of member function 'std::__cxx11::basic_string<char, std::char_traits<char>, glslang::pool_allocator<char> >::swap' requested here
        this->swap(__str);
              ^
/workspace/glslang/glslang/MachineIndependent/Initialize.cpp:3854:46: note: in instantiation of member function 'std::__cxx11::basic_string<char, std::char_traits<char>, glslang::pool_allocator<char> >::operator=' requested here
                                    typeName = sampler.getString();
                                             ^
/workspace/glslang/glslang/MachineIndependent/../Include/PoolAlloc.h:318:21: note: declared protected here
    pool_allocator& operator=(const pool_allocator&) { return *this; }
                    ^
In file included from /workspace/glslang/glslang/MachineIndependent/Initialize.cpp:52:
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/intermediate.h:54:
In file included from /workspace/glslang/glslang/MachineIndependent/../Include/Common.h:55:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/sstream:38:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/istream:38:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/ios:40:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/char_traits.h:39:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_algobase.h:64:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/stl_pair.h:59:
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h:187:11: error: 'operator=' is a protected member of 'glslang::pool_allocator<char>'
      __b = _GLIBCXX_MOVE(__tmp);
      ~~~ ^ ~~~~~~~~~~~~~~~~~~~~
/workspace/glslang/glslang/MachineIndependent/../Include/PoolAlloc.h:318:21: note: declared protected here
    pool_allocator& operator=(const pool_allocator&) { return *this; }
                    ^
2 errors generated.
make[2]: *** [glslang/CMakeFiles/glslang.dir/build.make:159: glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o] Error 1

```

**Comment by johnkslang** [comment]
Is this the same code that is passing the gcc and clang bot tests?

**Comment by johnkslang** [comment]
I suspect it is because a TString was overwritten, as that's the only real functional change. I modified the code to be simpler, and added some related `const`.

**Comment by EoD** [comment]
Commit 12d693682f8a5080c2b4a9347dfdaea66e7b4de7 fixes the error.

I don't really understand why it complains about the ``operator=`` in the first hand, but as you added ``const``, do you want to change [Line 3854]( https://github.com/KhronosGroup/glslang/blob/12d693682f8a5080c2b4a9347dfdaea66e7b4de7/glslang/MachineIndependent/Initialize.cpp#L3854) to ``const TString`` as well?

**Comment by johnkslang** [comment]
Yes, I think it would be better to also use `const TString` (thought of that after pushing). But, as a first step it is good to know that was the problem: must be the assignment of an already allocated `char` in the string to the new character, and I guess that is not done elsewhere.  So many compilers are accepting it the way it was.

**Comment by EoD** [comment]
Does your compiler not parse [Line 3854 in Initialize.cpp](https://github.com/KhronosGroup/glslang/blob/12d693682f8a5080c2b4a9347dfdaea66e7b4de7/glslang/MachineIndependent/Initialize.cpp#L3854) by using the protected ``operator=`` from [Line 318 in PoolAlloc.h](https://github.com/KhronosGroup/glslang/blob/32fd5d26e3576619205671f12d125d6375252fc6/glslang/Include/PoolAlloc.h#L318)? Or does it just ignore the ``protected``?

**Comment by johnkslang** [comment]
> Does your compiler not parse Line 3854 in Initialize.cpp by using the protected operator= from Line 318 in PoolAlloc.h?  Or does it just ignore the protected?

Apparently, this is implemented some other way, in lots of compilers.

> EDIT: Might this actually be a MSVC bug which allows the code to compile?

Keep in mind, this passed all the clang, gcc, and Windows bot tests, compiling both debug and release. This includes compiling and passing a test that would not run pass without executing the line of code in question.  I wonder if it is sensitive to compiler version, or recent change in the C++ standard on how to handle assignment to an existing std::string.

**Comment by EoD** [comment]
I went ahead and tested the previous errors on an Intel dev machine (using the icc). It shows the same problem:
```
[ 28%] Building CXX object glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h(186): error #373: "glslang::pool_allocator<T> &glslang::pool_allocator<T>::operator=(const glslang::pool_allocator<T> &) [with T=char]" (declared at line 318 of "/workspace/glslang/glslang/MachineIndependent/../Include/../Include/PoolAlloc.h") is inaccessible
        __a = _GLIBCXX_MOVE(__b);
            ^
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/move.h(187): error #373: "glslang::pool_allocator<T> &glslang::pool_allocator<T>::operator=(const glslang::pool_allocator<T> &) [with T=char]" (declared at line 318 of "/workspace/glslang/glslang/MachineIndependent/../Include/../Include/PoolAlloc.h") is inaccessible
        __b = _GLIBCXX_MOVE(__tmp);
            ^
compilation aborted for /workspace/glslang/glslang/MachineIndependent/Initialize.cpp (code 2)
make[2]: *** [glslang/CMakeFiles/glslang.dir/build.make:159: glslang/CMakeFiles/glslang.dir/MachineIndependent/Initialize.cpp.o] Error 2
make[1]: *** [CMakeFiles/Makefile2:1004: glslang/CMakeFiles/glslang.dir/all] Error 2
make: *** [Makefile:139: all] Error 2
```

I am not sure how this passed your bot test, as it fails on 3 different compilers on two different machines for me. Are you using cmake as well? Is the test identical to the documented build process in the README?

**Comment by johnkslang** [comment]
Agreed it is mysterious.  [See this log](https://travis-ci.org/KhronosGroup/glslang/builds/229855943?utm_source=github_status&utm_medium=notification) for 6 configurations passing.  Everything should be using CMake.  Maybe flag-setting difference?

**Comment by EoD** [comment]
The major difference is the version of GCC you are using is 4.8, while all the systems I tested had 5.4.0 installed. I am not able to reproduce the bug on either 4.9.4 nor on 6.3.0, only 5.4.0 seems to be affected.

I am not an expert, but as far as I can tell, both clang and icc are using the GCC tools which are installed on the system while compiling, so this explains the consistent failures there: https://software.intel.com/en-us/node/

I tried looking for a known regression, but so far I have been unlucky.

**Comment by EoD** [comment]
Apparently this is a bug in the GCC headers and is fixed in GCC 5.5.0+:
https://gcc.gnu.org/PR58265
https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=5464a27a8faed210695ce5a4ff1133bc415f0d54

After applying the patch to GCC, I was not able to reproduce the crash before.

Sorry for the noise!

**Comment by johnkslang** [comment]
Thanks for figuring it out, @EoD.
